lib/grammers-client/examples/downloader.rs

```diff
@@ -0,0 +1,195 @@
+                );
+                sign_out = true;
+            }
+        }
+    }
+
+    // Obtain a `ClientHandle` to perform remote calls while `Client` drives the connection.
+    //
+    // This handle can be `clone()`'d around and freely moved into other tasks, so you can invoke
+    // methods concurrently if you need to. While you do this, the single owned `client` is the
+    // one that communicates with the network.
+    //
+    // The design's annoying to use for trivial sequential tasks, but is otherwise scalable.
+    let mut client_handle = client.clone();
+    let network_handle = task::spawn(async move { client.run_until_disconnected().await });
+
+    let chat = client_handle.resolve_username(chat_name.as_str()).await?;
+
+    if chat.is_none() {
+        panic!("Chat {} could not be found", chat_name);
+    }
+
+    let mut messages = client_handle.iter_messages(&chat.unwrap());
+
+    println!(
+        "Chat {} has {} total messages.",
+        chat_name,
+        messages.total().await.unwrap()
+    );
+
+    let mut counter = 0;
+    loop {
+        match messages.next().await {
+            Ok(Some(msg)) => {
+                counter = counter + 1;
+                println!("Message {}:{}", msg.id(), msg.text());
+                if let Some(media) = msg.media() {
+                    let dest = format!(
+                        "target/message-{}{}",
+                        &msg.id().to_string(),
+                        get_file_extension(&media)
+                    );
+                    let path = Path::new(dest.as_str());
+                    client_handle
+                        .download_media(&media, &path)
+                        .await
+                        .expect("Error downloading message");
+                }
+            }
+            Err(e) => panic!("{}", e),
+            Ok(None) => break { println!("Finished iterating messages") },
+        }
+    }
+
+    println!("Downloaded {} messages", counter);
+
+    if sign_out {
+        // TODO revisit examples and get rid of "handle references" (also, this panics)
+        drop(client_handle.sign_out_disconnect().await);
+    }
+
+    network_handle.await??;
+    Ok(())
+}
+
+fn main() -> Result<()> {
+    runtime::Builder::new_current_thread()
+        .enable_all()
+        .build()
+        .unwrap()
+        .block_on(async_main())
+}
+
+fn get_file_extension(media: &Media) -> String {
+    let suffix = match media {
+        Photo(_) => "jpg".to_string(),
+        Sticker(sticker) => get_subtype(sticker.document.mime_type()),
+        Document(document) => get_subtype(document.mime_type()),
+        Contact(_) => "vcf".to_string(),
+        _ => String::new(),
+    };
+    return format!(".{}", suffix);
+}
+
+fn get_subtype(mime_type: Option<&str>) -> String {
+    return mime_type
+        .map(|m| {
+            let mime: Mime = m.parse().unwrap();
+            mime.subtype().to_string()
+        })
+        .unwrap_or(String::new());
```

Comment:
This might create files such as `message-1001.` (empty extension). Perhaps it would be better to use the extension from the media's file name?

Reply:
wdym, it will always try to use the mime type if it's present, otherwise it is not possible to know the extension unless inspecting the bytes of the content

Reply:
I meant there would be a trailing dot (`message-1001.` rather than `message-1001`).

Reply:
Fixed in https://github.com/Lonami/grammers/pull/148